language/move-bytecode-verifier/src/control_flow.rs

```diff
@@ -7,268 +7,82 @@
-    }
-
-    fn labeled_code<'b: 'a>(
-        &self,
-        labels: &'b [Label],
-    ) -> impl Iterator<Item = (CodeOffset, &'a Bytecode, &'b Label)> {
-        self.code()
-            .zip(labels)
-            .map(|((i, instr), lbl)| (i, instr, lbl))
-    }
-
-    fn error(&self, status: StatusCode, offset: CodeOffset) -> PartialVMError {
-        PartialVMError::new(status).at_code_offset(self.current_function, offset)
-    }
-}
-
-fn instruction_labels(context: &ControlFlowVerifier) -> Vec<Label> {
-    let mut labels: Vec<Label> = (0..context.code.len()).map(|_| Label::Code).collect();
-    let mut loop_continue = |loop_idx: CodeOffset, last_continue: CodeOffset| {
-        labels[loop_idx as usize] = Label::Loop { last_continue }
-    };
-    for (i, instr) in context.code() {
-        match instr {
-            // Back jump/"continue"
-            Bytecode::Branch(prev) | Bytecode::BrTrue(prev) | Bytecode::BrFalse(prev)
-                if is_back_edge(i, *prev) =>
-            {
-                loop_continue(*prev, i)
-            }
-            _ => (),
-        }
+        Some(_) => Ok(()),
     }
-    labels
 }
 
-// Ensures the invariant:
-//   - All forward jumps do not enter into the middle of a loop
-//   - All "breaks" go to the "end" of the loop
-//   - All back jumps are only to the current loop
-//   - Nested loops do not exceed a given depth
-fn check_jumps(
+pub fn verify_reducibility<'a>(
     verifier_config: &VerifierConfig,
-    context: &ControlFlowVerifier,
-    labels: Vec<Label>,
+    function_view: &'a FunctionView<'a>,
 ) -> PartialVMResult<()> {
-    // All back jumps are only to the current loop
-    check_continues(context, &labels)?;
-    // All "breaks" go to the "end" of the loop
-    check_breaks(context, &labels)?;
-
-    let loop_depth = count_loop_depth(&labels);
+    let current_function = function_view.index().unwrap_or(FunctionDefinitionIndex(0));
+    let err = move |code: StatusCode, offset: CodeOffset| {
+        Err(PartialVMError::new(code).at_code_offset(current_function, offset))
+    };
 
-    // All forward jumps do not enter into the middle of a loop
-    check_no_loop_splits(context, &labels, &loop_depth)?;
-    // Nested loops do not exceed a given depth
-    check_loop_depth(verifier_config, context, &labels, &loop_depth)
-}
+    let summary = LoopSummary::new(function_view.cfg());
+    let mut partition = LoopPartition::new(&summary);
 
-fn check_code<
-    F: FnMut(&Vec<(CodeOffset, CodeOffset)>, CodeOffset, &Bytecode) -> PartialVMResult<()>,
->(
-    context: &ControlFlowVerifier,
-    labels: &[Label],
-    mut check: F,
-) -> PartialVMResult<()> {
-    let mut loop_stack: Vec<(CodeOffset, CodeOffset)> = vec![];
-    for (cur_instr, instr, label) in context.labeled_code(labels) {
-        // Add loop to stack
-        if let Label::Loop { last_continue } = label {
-            loop_stack.push((cur_instr, *last_continue));
+    for head in summary.preorder().rev() {
```

Comment:
Some extensive comments here would really help me. Took me about an hour to figure out what was going on lol

Reply:
The code is very clean! But a bit dense, if you get me